File: pushnotification/src/main/java/io/castled/notifications/logger/CastledLogger.java
package io.castled.notifications.logger;

import android.util.Log;

import androidx.annotation.NonNull;

import io.castled.notifications.Logger;

public class CastledLogger implements Logger {

    public static final String TAG = "castled";
    private static final CastledLogger logger = new CastledLogger();

    @Override
    public void verbose(@NonNull String message) {
        Log.v(TAG, message);
    }

    @Override
    public void debug(@NonNull String message) {
        Log.d(TAG, message);
    }

    @Override
    public void info(@NonNull String message) {
        Log.i(TAG, message);
    }

    @Override
    public void warning(@NonNull String message) {
        Log.w(TAG, message);
    }

    @Override
    public void error(@NonNull String message) {
        Log.e(TAG, message);
    }

    @Override
    public void error(@NonNull String message, @NonNull Throwable throwable) {
        Log.v(TAG, message, throwable);
    }

    public static CastledLogger getInstance() {
        return logger;
    }
}

File: castled-notifications/src/main/java/io/castled/android/notifications/logger/CastledLogger.kt
package io.castled.android.notifications.logger

import android.util.Log

class CastledLogger private constructor(private val tag: String) : Logger {
    override fun verbose(message: String) {
        Log.v(tag, message)
    }

    override fun debug(message: String) {
        Log.d(tag, message)
    }

    override fun info(message: String) {
        Log.i(tag, message)
    }

    override fun warning(message: String) {
        Log.w(tag, message)
    }

    override fun error(message: String) {
        Log.e(tag, message)
    }

    override fun error(message: String, throwable: Throwable) {
        Log.e(tag, message, throwable)
    }

    companion object {

        private val loggersByTag = mutableMapOf<String, CastledLogger>()

        @Synchronized
        fun getInstance(tag: String): CastledLogger {
            return loggersByTag.getOrPut(tag) {
                CastledLogger(tag)
            }
        }
    }
}